src/components/OpenWith/OpenWith.js

```diff
@@ -112,16 +121,60 @@ class OpenWith extends PureComponent<Props, State> {
      * @return {void}
      */
     componentDidMount() {
-        const { fileId }: Props = this.props;
+        this.fetchOpenWithData();
+    }
+
+    /**
+     * Fetches file extension and Open With data.
+     *
+     * @return {void}
+     */
+    fetchOpenWithData() {
+        const { fileId, extension, language }: Props = this.props;
+        if (extension) {
+            this.setState({ extension });
+        } else {
+            this.api
+                .getFileAPI()
+                .getFile(fileId, this.fetchExtensionSuccessCallback, noop, {
+                    fields: [FIELD_EXTENSION],
+                });
+        }
+
         this.api
             .getOpenWithAPI(false)
             .getOpenWithIntegrations(
                 fileId,
+                language,
                 this.fetchOpenWithSuccessHandler,
                 this.fetchErrorCallback,
             );
     }
 
+    /**
+     * Handles a successful files call to get the file extension.
+     *
+     * @param {object} extensionData - The files call response with the file extension
+     * @return {void}
+     */
+    fetchExtensionSuccessCallback = ({ extension }: { extension: string }) => {
+        this.setState({ extension });
+    };
+
+    /**
+     * After component updates, re-fetch Open With data if appropriate.
+     *
+     * @return {void}
+     */
+    componentDidUpdate(prevProps: Props): void {
+        const { fileId: currentFileId }: Props = this.props;
+        const { fileId: previousFileId }: Props = prevProps;
+
+        if (currentFileId !== previousFileId) {
```

Comment:
Would we only want to do this if the extension changes?

Reply:
If the file ID changes, then we need to get new open with data and a new extension. Either that will be provided via props or we'll fetch for it.

Reply:
Ah, okay. I didn't realize the method was doing two different things. It seems like it should still be two methods to me, but it's not really a big deal.